source/postprocess/topography.cc

```diff
@@ -101,57 +83,196 @@ namespace aspect
+        }
+
+      // Just return stats if text output is not required at all or not needed at this time
+      if (!write_to_file || ((this->get_time() < last_output_time + output_interval)
+                             && (this->get_timestep_number() != 0)))
+        return std::pair<std::string, std::string> ("Topography min/max:",
+                                                    output_stats.str());
+
+      std::string filename = this->get_output_directory() +
+                             "topography." +
+                             Utilities::int_to_string(this->get_timestep_number(), 5);
+      if (this->get_parameters().run_postprocessors_on_nonlinear_iterations)
+        filename.append("." + Utilities::int_to_string (this->get_nonlinear_iteration(), 4));
+
+      const unsigned int max_data_length = Utilities::MPI::max (output_file.str().size()+1,
+                                                                this->get_mpi_communicator());
+      const unsigned int mpi_tag = 777;
+
+      // on processor 0, collect all of the data the individual processors send
+      // and concatenate them into one file
+      if (Utilities::MPI::this_mpi_process(this->get_mpi_communicator()) == 0)
+        {
+          std::ofstream file (filename.c_str());
+
+          file << "# "
+               << ((dim==2)? "x y" : "x y z")
+               << " topography" << std::endl;
+
+          // first write out the data we have created locally
+          file << output_file.str();
+
+          std::string tmp;
+          tmp.resize (max_data_length, '\0');
+
+          // then loop through all of the other processors and collect
+          // data, then write it to the file
+          for (unsigned int p=1; p<Utilities::MPI::n_mpi_processes(this->get_mpi_communicator()); ++p)
+            {
+              MPI_Status status;
+              // get the data. note that MPI says that an MPI_Recv may receive
+              // less data than the length specified here. since we have already
+              // determined the maximal message length, we use this feature here
+              // rather than trying to find out the exact message length with
+              // a call to MPI_Probe.
+              MPI_Recv (&tmp[0], max_data_length, MPI_CHAR, p, mpi_tag,
+                        this->get_mpi_communicator(), &status);
+
+              // output the string. note that 'tmp' has length max_data_length,
+              // but we only wrote a certain piece of it in the MPI_Recv, ended
+              // by a \0 character. write only this part by outputting it as a
+              // C string object, rather than as a std::string
+              file << tmp.c_str();
+            }
+        }
+      else
+        // on other processors, send the data to processor zero. include the \0
+        // character at the end of the string
+        {
+          MPI_Send (&output_file.str()[0], output_file.str().size()+1, MPI_CHAR, 0, mpi_tag,
```

Comment:
you will get an out-of-bounds error here because you assume that `output_file.str()` points to a string that is terminated by a `\0` character that `str().size()` does not actually count. This may be the case, or maybe not -- I don't think that you can count on that. So I think you really should only transmit `str().size()` characters. On the receiving side, you will then have to make sure you deal with the fact that no `\0` was transmitted -- or you append a `\0` character before sending and then you don't have to deal with it at the receiving end.